test/extended/router/stress.go

```diff
@@ -270,34 +212,162 @@ var _ = g.Describe("[sig-network][Feature:Router][apigroup:route.openshift.io]",
 			o.Expect(err).NotTo(o.HaveOccurred())
 
 			g.By("verifying that only get a few updates")
-			writes = 0
-			w, err = client.Watch(context.Background(), metav1.ListOptions{Watch: true, ResourceVersion: route.ResourceVersion})
+			verifyConflictingWrites(client, route.ResourceVersion, 10*time.Second, 5, "9")
+		})
+
+		g.It("converges when multiple routers are writing conflicting upgrade validation status", func() {
+			g.By("deploying a scaled out namespace scoped router that adds the UnservableInFutureVersions condition")
+
+			rsAdd, err := oc.KubeClient().AppsV1().ReplicaSets(ns).Create(
+				context.Background(),
+				scaledRouter(
+					"router-add-condition",
+					routerImage,
+					[]string{
+						"-v=5",
+						fmt.Sprintf("--namespace=%s", ns),
+						// the contention tracker is resync / 10, so this will give us 2 minutes of contention tracking
+						"--resync-interval=20m",
+						"--name=conflicting",
+						"--debug-upgrade-validation-force-add-condition",
+					},
+				),
+				metav1.CreateOptions{},
+			)
 			o.Expect(err).NotTo(o.HaveOccurred())
-			func() {
-				defer w.Stop()
-				timer := time.NewTimer(10 * time.Second)
-				ch := w.ResultChan()
-			Wait:
-				for i := 0; ; i++ {
-					select {
-					case obj, ok := <-ch:
-						o.Expect(ok).To(o.BeTrue())
-						if r, ok := obj.Object.(*routev1.Route); ok {
-							if r == nil || r.Name != "9" {
-								continue
-							}
+			err = waitForReadyReplicaSet(oc.KubeClient(), ns, rsAdd.Name)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("creating multiple routes")
+			client := routeclientset.NewForConfigOrDie(oc.AdminConfig()).RouteV1().Routes(ns)
+			var rv string
+			createTestRoutes(client, 20)
+
+			g.By("waiting for sufficient routes to have a UnservableInFutureVersions status condition")
+			err = wait.Poll(time.Second, 2*time.Minute, func() (bool, error) {
+				routes, err := client.List(context.Background(), metav1.ListOptions{})
+				if err != nil {
+					return false, err
+				}
+				o.Expect(routes.Items).To(o.HaveLen(20))
```

Comment:
Won't this mark any failure to match 20? You're looping for 2 mins - shouldn't we check for an error from the poll loop instead of asserting here?